modules/custom-post-types/redirector.php

```diff
@@ -0,0 +1,211 @@
+
+		if ( empty( $request_path ) ) {
+			return;
+		}
+
+		$redirect_uri = self::get_redirect_uri( $request_path );
+
+		if ( empty( $redirect_uri ) ) {
+			return;
+		}
+
+		header( 'X-jetpack-redirector: HIT' );
+		$redirect_status = apply_filters( 'jetpack_redirector_redirect_status', 301, $url );
+		wp_safe_redirect( $redirect_uri, $redirect_status );
+		exit;
+	}
+
+	static function get_redirect_uri( $passed_url ) {
+		$url = self::normalize_url( $passed_url );
+
+		if ( is_wp_error( $url ) ) {
+			return false;
+		}
+
+		// Allowed list of Params that should be pass through as is.
+		$protected_params = apply_filters( 'jetpack_redirector_preserve_query_params', array(), $url );
+		$protected_param_values = array();
+		$param_values = array();
+
+		// Parse URL to get Query Params.
+		$query_params = wp_parse_url( $url, PHP_URL_QUERY );
+
+		if ( ! empty( $query_params ) ) {
+			// Parse Query String to Associated Array.
+			parse_str( $query_params, $param_values );
+
+			// For every allowed param save value and strip from url
+			foreach ( $protected_params as $protected_param ) {
+				if ( ! empty( $param_values[ $protected_param ] ) ) {
+					$protected_param_values[ $protected_param ] = $param_values[ $protected_param ];
+					$url = remove_query_arg( $protected_param, $url );
+				}
+			}
+		}
+
+		$url_hash = self::get_url_hash( $url );
+
+		$redirect_post_id = wp_cache_get( $url_hash, self::CACHE_GROUP );
+
+		if ( false === $redirect_post_id ) {
+			$redirect_post_id = self::get_redirect_post_id( $url );
+			wp_cache_add( $url_hash, $redirect_post_id, self::CACHE_GROUP );
+		}
+
+		if ( ! $redirect_post_id ) {
+			return false;
+		}
+
+		$redirect_post = get_post( $redirect_post_id );
+
+		if ( ! $redirect_post instanceof WP_Post ) {
+			// If redirect post object doesn't exist, update the cache
+			wp_cache_set( $url_hash, 0, self::CACHE_GROUP );
+			return false;
+		}
+
+		if ( 0 !== $redirect_post->post_parent ) {
+			$parent_post = get_post( $redirect_post->post_parent );
+			if ( ! $parent_post instanceof WP_Post ) {
+				return false;
+			}
+
+			// Support redirecting directly to attached files...
+			if ( $parent_post->post_type === 'attachment' ) {
+				$ext_info = wp_check_filetype( $url );
+				// ...but only if the extension is allowed & matches
+				if ( $ext_info['ext'] && preg_match( '/\.' . $ext_info['ext'] . '$/', $parent_post->guid ) ) {
+					return $parent_post->guid;
+				}
+			}
+
+			// Add allowed params to the redirect URL.
+			return add_query_arg( $protected_param_values, get_permalink( $parent_post ) );
+		}
+
+		if ( ! empty( $redirect_post->post_excerpt ) ) {
+			// Add allowed params to the redirect URL.
+			return add_query_arg( $protected_param_values, esc_url_raw( $redirect_post->post_excerpt ) );
+		}
+	}
+
+	static function get_redirect_post_id( $url ) {
+		global $wpdb;
+
+		$redirect_post_id = $wpdb->get_var( $wpdb->prepare(
```

Comment:
Should we use a `WP_Query` here instead? It would add some amount caching, but this query is indexed & super fast as-is.